BSB_LAN/BSB_LAN.ino

```diff
@@ -6177,30 +6312,128 @@ void loop() {
+                n = sscanf(p+2, "%d-%d-%d,%d-%d-%d", &ay,&am,&ad, &by,&bm,&bd);
+                if (n<1 || ay<1) transmitFile(dataFile);  // no or zero limit?
+                else { // limited datalog requested
+                  // transfer header (no error message in case of read() error => handled in subsequent reading of the actual data, anyway):
+                  client.write((byte*)datalogFileHeader,strlen(datalogFileHeader));
+                  File indexFile = SD.open(datalogIndexFileName);
+                  if (!indexFile) transmitFile(dataFile);  // no index??
+                  else {
+                    unsigned long datalogTargetPosition;
+                    if (n==6) { // date range requested
+                      compactDate_t a, b, date, prevDate;
+                      a.elements.year = ay;
+                      a.elements.month = am;
+                      a.elements.day = ad;
+                      b.elements.year = by;
+                      b.elements.month = bm;
+                      b.elements.day = bd;
+                      prevDate.combined = 0;
+                      unsigned long datalogFromPosition=0, datalogToPosition=0;
+                      // read index file backwards (why? => https://github.com/fredlcore/BSB-LAN/issues/539#issuecomment-1464885147):
+                      for (unsigned long pos=indexFile.size(); pos>=datalogIndexEntrySize; pos-=datalogIndexEntrySize) {
+                        indexFile.seek(pos-datalogIndexEntrySize);
+                        if (indexFile.read((byte*)&date,sizeof(date)) <= 0 ||
+                            indexFile.read((byte*)&datalogTargetPosition,sizeof(datalogTargetPosition)) <= 0)
+                          break; // nothing could be read!
+                        if (prevDate.combined && prevDate.combined < date.combined) break; // dates running backwards?
+                        if (date.combined >  b.combined) datalogToPosition = datalogTargetPosition;
+                        if (date.combined >= a.combined) datalogFromPosition = datalogTargetPosition;
+                        else break; // we're already before date a in the index
+                        prevDate = date;
+                      }
+                      if (datalogFromPosition) {
+                        dataFile.seek(datalogFromPosition);
+                        if (!datalogToPosition) datalogToPosition = dataFile.size();
+                        unsigned long nBytesToDo = datalogToPosition - datalogFromPosition;
+                        // the following re-uses code fragments from transmitFile():
+                        unsigned logbuflen = (OUTBUF_USEFUL_LEN + OUTBUF_LEN > 1024)?1024:(OUTBUF_USEFUL_LEN + OUTBUF_LEN);
+                        byte *buf = 0;
+#ifdef ESP32 // Arduino seems to have problems with the bigger, malloc'ed buffer
+                        buf = (byte*)malloc(4<<10);  // try to use 4 KB buffer, for improved transfer rates
```

Comment:
Why would anyone ever do this? `4<<10` isn't saving space compared to 4096, and the compiler will treat it both the same. This will just force me to look at PRs even more closely and waste both of our time. Just now because I was looking for the 4096 byte buffer and only found it after going through the diff line by line. Grrr...

Reply:
Sorry, but to me 1<<10 is clearly 1 KB. I could also easily recognize 4096, but for 7*1024 I would grab a calculator, whereas 7<<10 is easier for me to get right. However, I'll change 4<<10 to 4096 for you, if you wish so.

Reply:
> Please give it another try with the most recent version of this PR here. If that works, I'll also apply [4c5f7a7](https://github.com/fredlcore/BSB-LAN/commit/4c5f7a7864f267fe776fb3e3465895f9b8182e6a) to #554.

I had quoted https://github.com/fredlcore/BSB-LAN/commit/4c5f7a7864f267fe776fb3e3465895f9b8182e6a here to save you time in localizing my most recent change.

Reply:
> Sorry, but to me 1<<10 is clearly 1 KB. I could also easily recognize 4096, but for 7*1024 I would grab a calculator, whereas 7<<10 is easier for me to get right. However, I'll change 4<<10 to 4096 for you, if you wish so.

Also, I would have to double check whether 1048576 is 1 MB. With 1<<20, I wouldn't have that problem.
This just to answer your "Why would anyone ever do this?"

Reply:
Even if it was necessary to exactly use 4096 bytes ((which is not the case here, so 4000 would have been as fine), _I_ would take a look at the surrounding code and see that this style of writing isn't used here. And three lines above, there's 1024 and not 2<<8, so _I_'d take a hint.

Reply:
Hint taken, will change this.

Reply:
👍

Reply:
> Even if it was necessary to exactly use 4096 bytes ((which is not the case here, so 4000 would have been as fine), _I_ would take a look at the surrounding code and see that this style of writing isn't used here. And three lines above, there's 1024 and not 2<<8, so _I_'d take a hint.

4096 instead of 4000 has been chosen intentionally. Afaik are SDs and SPIFFS block devices, byte aligned in powers of two. Depending on what buffering SD.read() does (if any), this might just make a difference when reading a file right from its first byte, as in /D. However, I have not noticed a discernible difference between /D and /Dn or /Da,b (where I didn't go to the trouble of aligning reads relative to the file's start) so I guess it really didn't matter here.